File: Pollard_Rho_Algorithm/Javascript/jcla1/pollard.js
function defaultPollard(n) {
    return pollard(n, function(x) {
        return (x*x - 1) % n;
    });
}

// Takes f(x, n) where n is the integer to factorize
function pollard(n, f) {
    var x = 2,
        y = 2,
        d = 1;

    while (d == 1) {
        x = f(x, n);
        y = f(f(y, n), n);
        d = gcd(Math.abs(x - y), n);
    }

    if  (d == n) return -1;
    return d;
}

function gcd(a, b) {
    if (b == 0) return a;
    return gcd(b, a % b);
}

File: Pollard_Rho_Algorithm/Go/jcla1/pollard.go
package pollard

import (
	"math"
)

func Pollard(n int, f func(int, int) int) int {
	x, y, d := 2, 2, 1
	for d == 1 {
		x = f(x, n)
		y = f(f(x, n), n)
		d = gcd(int(math.Abs(float64(x-y))), n)
	}

	if n == d {
		return -1
	}

	return d
}

func gcd(a, b int) int {
	if b == 0 {
		return a
	}
	return gcd(b, a%b)
}
